Packages/com.unity.inputsystem/InputSystem/Plugins/DualShock/MoveControllerPS4.cs

```diff
@@ -0,0 +1,380 @@
+                throw new ArgumentException("Slot index out of range: " + slotIndex, "slotIndex");
+
+            if (moveIndex < 0 || moveIndex >= 2)
+                throw new ArgumentException("Move index out of range: " + slotIndex, "moveIndex");
+
+            int realIndex = (slotIndex * 2) + moveIndex;
+
+            if (s_Devices[realIndex] != null && s_Devices[realIndex].slotIndex == slotIndex)
+            {
+                return s_Devices[realIndex];
+            }
+
+            return null;
+        }
+
+        protected override void OnAdded()
+        {
+            base.OnAdded();
+
+            AddDeviceToList();
+        }
+
+        protected override void OnRemoved()
+        {
+            base.OnRemoved();
+
+            RemoveDeviceFromList();
+        }
+
+        private void AddDeviceToList()
+        {
+            int realIndex = (slotIndex * 2) + moveIndex;
+
+            if (realIndex >= 0 && realIndex < s_Devices.Length)
+            {
+#if !UNITY_EDITOR
+                Debug.Assert(s_Devices[realIndex] == null, "PS4 move controller with same slotIndex/moveIndex already added");
+#endif
+                s_Devices[realIndex] = this;
+            }
+        }
+
+        private void RemoveDeviceFromList()
+        {
+            int realIndex = (slotIndex * 2) + moveIndex;
+
+            if (realIndex >= 0 && realIndex < s_Devices.Length && s_Devices[realIndex] == this)
+                s_Devices[realIndex] = null;
+        }
+
+        protected override void FinishSetup(InputDeviceBuilder builder)
+        {
+            selectButton = builder.GetControl<ButtonControl>(this, "select");
+            triggerButton = builder.GetControl<ButtonControl>(this, "triggerButton");
+            moveButton = builder.GetControl<ButtonControl>(this, "move");
+            startButton = builder.GetControl<ButtonControl>(this, "start");
+
+            squareButton = builder.GetControl<ButtonControl>(this, "square");
+            triangleButton = builder.GetControl<ButtonControl>(this, "triangle");
+            circleButton = builder.GetControl<ButtonControl>(this, "circle");
+            crossButton = builder.GetControl<ButtonControl>(this, "cross");
+
+            trigger = builder.GetControl<ButtonControl>(this, "trigger");
+
+            accelerometer = builder.GetControl<Vector3Control>(this, "accelerometer");
+            gyro = builder.GetControl<Vector3Control>(this, "gyro");
+
+            base.FinishSetup(builder);
+
+            var capabilities = description.capabilities;
+            var deviceDescriptor = PS4MoveDeviceDescriptor.FromJson(capabilities);
+
+            if (deviceDescriptor != null)
+            {
+                m_SlotId = (int)deviceDescriptor.slotId;
+                m_MoveIndex = (int)deviceDescriptor.moveIndex;
+                m_DefaultColorId = (int)deviceDescriptor.defaultColorId;
+                m_PS4UserId = (int)deviceDescriptor.userId;
+            }

```

Comment:
I think for now this is fine. There's a change that has happened recently which I think will require us to more cleanly separate certain pieces of device setup/configuration. In order to not create garbage when the user unplugs devices and plugs them back in, the system now retains devices that have been reported by the runtime and have been removed through device remove events. So we may end up re-using InputDevice instances in certain cases.

Right now, the descriptor actually has to match 100% so there shouldn't be a problem in this particular case.

At some point, I think we probably want to make sure that all one-time configuration on devices gets reset when they are removed.